### PR TITLE
[Breaking] feat: self-redeem

### PR DESCRIPTION
**Pull request by sander2**
(User story copied from notion)

> ### User Story
> 
> As a Vault
> 
> I want to instantly redeem iBTC or kBTC
> 
> So that I 
> 
> 1. Can reuse my collateral atomically
> 2. Don’t have to send a BTC transaction to myself wasting BTC tx fees
> 3. And have the BTC in my wallet available right away (technically already the case since we don’t have theft reporting anymore)
> 
> ### Background
> 
> When vaults redeem with themselves, they basically burn iBTC, unlock BTC locked with them, and free collateral. This is a three step process at the moment:
> 
> - request_redeem schedules the redeem to be processed and gives the vault a BTC address from its own wallet to transfer BTC from its other addresses
> - the vault client listens for this request_redeem event and then sends the BTC to the specified address (but all in the same wallet - it’s just shifting UTXOs around)
> - once the BTC transaction has enough confirmations, the vault client calls execute_redeem which burns the locked iBTC and frees the vault collateral
> 
> This process takes at least 1 hour on main net due to the 6 required BTC confirmations.
> 
> ### Advantages
> 
> By making redeem atomic for vaults, the time to execute redeems is reduced to 12-36 seconds (blocks until the parachains is finalized).
> 
> Collateral for vaults is directly unlocked allowing vaults to:
> 
> 1. Instantly increase their collateral rate in case they need to (not in 1 hour)
> 2. Atomically execute actions with their collateral.

Note that the user still pays the 0.5% redeem fees unless it redeems the full amount left in the vault

### CODE REVIEW

File: crates/redeem/src/lib.rs [hunk 1]
```diff
@@ -336,9 +341,103 @@ pub mod pallet {
...
+        update_vault_tokens::<T>(&vault_id, &consumed_issued_tokens)?;
+
+        Pallet::<T>::deposit_event(Event::<T>::SelfRedeem {
+            vault_id,
+            amount: consumed_issued_tokens.amount(),
+            fee: fees.amount(),
+        });
+
+        Ok(())
+    }
```
Comment (sander2) [inline crates/redeem/src/lib.rs:386]: My original code looked surprisingly complex for such a simple function, so this was a bit of an experiment to attempt to make the code easier to read. I like how clear this function is now, although I did end up paying the price in increased verbosity.

File: crates/redeem/src/lib.rs [hunk 2]
```diff
@@ -336,9 +341,103 @@ pub mod pallet {
...
+    use super::*;
+
+    pub(crate) fn execute<T: Config>(vault_id: DefaultVaultId<T>, amount_wrapped: Amount<T>) -> DispatchResult {
+        // ensure that vault is not liquidated and not banned
+        ext::vault_registry::ensure_not_banned::<T>(&vault_id)?;
+
+        // for self-redeem, dustAmount is effectively 1 satoshi
+        ensure!(!amount_wrapped.is_zero(), Error::<T>::AmountBelowDustAmount);
+
+        let (fees, consumed_issued_tokens) = calculate_token_amounts::<T>(&vault_id, &amount_wrapped)?;
```
Comment (gregdhill) [inline crates/redeem/src/lib.rs:373]: A potentially nice refactoring from a concept I came across whilst digging into Bitcoin Core is to bundle this tuple into something similar to the `COutput` and then calculate the amount to burn using [`GetEffectiveValue`](https://github.com/bitcoin/bitcoin/blob/2bd9aa5a44b88c866c4d98f8a7bf7154049cba31/src/wallet/coinselection.h#L108).

File: crates/redeem/src/lib.rs [hunk 3]
```diff
@@ -336,9 +341,103 @@ pub mod pallet {
             Self::_mint_tokens_for_reimbursed_redeem(vault_id, redeem_id)?;
             Ok(().into())
         }
+
+        #[pallet::weight(<T as Config>::WeightInfo::self_redeem())]
+        #[transactional]
+        pub fn self_redeem(
```
Comment (gregdhill) [inline crates/redeem/src/lib.rs:347]: Should we bundle this into the existing redeem flow? I guess having a separate extrinsic is cleaner but not sure if there were other considerations in your design here

**Review by gregdhill — COMMENTED** [review]
No blocking concerns but may be worth adding some integration tests, what do you think @sander2?

**Comment by sander2** [comment]
> No blocking concerns but may be worth adding some integration tests, what do you think @sander2?

What do you mean? There _are_ integration tests, see `test_redeem.rs`?

File: crates/redeem/src/lib.rs (hunk 2 above)
Comment (sander2) [reply crates/redeem/src/lib.rs:373]: This function is only used within this module (and only once at that), so I don't think that refactoring is worth it

Comment (gregdhill) [reply crates/redeem/src/lib.rs:373]: Yeah just mentioning it because I thought it was pretty ;)

**Comment by gregdhill** [comment]
> > No blocking concerns but may be worth adding some integration tests, what do you think @sander2?
> 
> What do you mean? There _are_ integration tests, see `test_redeem.rs`?

Yes you are right, sorry I must have missed those - trying to do too many things at once

**Pull request merged**